### ISSUE TITLE
MPI failing on TitanXm with CUDA.jl v3.5

### ISSUE THREAD

**Issue by luraess** [eth-cscs/ImplicitGlobalGrid.jl#25]
Getting the following error running IGG with CUDA.jl v3.5 on Titan Xm GPUs with `Driver Version: 470.42.01    CUDA Version: 11.4` in the `update_halo` function. Using system CUDA install, and either Julia MPICH MPI artifact or system OpenMPI **without** CUDA-aware support.

Interestingly, this error does not occur:
- if using the same CUDA install with Tesla V100 GPUs (without CUDA-aware support)
- if using CUDA-aware MPI and setting (`export IGG_CUDAAWARE_MPI=1`) on both Titan Xm and Tesla V100

**Tmp Fix:** downgrading to CUDA.jl v3.3.6 👀 

```
ERROR: ERROR: LoadError: LoadError: ERROR: LoadError: ERROR: LoadError: CUDA error: invalid argument (code 1, ERROR_INVALID_VALUE)
Stacktrace:
  [1] CUDA error: invalid argument (code 1, ERROR_INVALID_VALUE)
Stacktrace:
  [1] throw_api_error(CUDA error: invalid argument (code 1, ERROR_INVALID_VALUE)
Stacktrace:
  [1] throw_api_error(res::res::CUDA error: invalid argument (code 1, ERROR_INVALID_VALUE)
Stacktrace:
  [1] throw_api_error(res::CUDA.cudaError_enum)
    @ CUDA.cudaError_enum)
    @ CUDA ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/throw_api_error(error.jl:91
  [2] macro expansion
    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/error.jl:101 [inlined]
  [3] cuMemHostGetDevicePointer_v2(pdptr::Base.RefValue{CUDA.CuPtr{Nothing}}, p::Ptr{Nothing}, Flags::Int64)
    @ CUDA ~/.julia/packages/CUDA/YpW0k/lib/utils/call.jl:26
  [4] convert
    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/memory.jl:130 [inlined]
  [5] unsafe_convert
    @ ~/.julia/packages/CUDA/YpW0k/src/array.jl:320 [inlined]
  [6] pointer
    @ ~/.julia/packages/CUDA/YpW0k/src/array.jl:275 [inlined]
  [7] unsafe_convert
    @ ~/.julia/packages/CUDA/YpW0k/src/array.jl:327 [inlined]
  [8] adapt_structure
    @ ~/.julia/packages/CUDA/YpW0k/src/compiler/execution.jl:139 [inlined]
  [9] adapt
    @ ~/.julia/packages/Adapt/RGNRk/src/Adapt.jl:40 [inlined]
 [10] cudaconvert(arg::CUDA.CuArray{Float64, 1, CUDA.Mem.HostBuffer})
    @ CUDA ~/.julia/packages/CUDA/YpW0k/src/compiler/execution.jl:152
 [11] map
    @ ./tuple.jl:216 [inlined]
 [12] macro expansion
    @ ~/.julia/packages/CUDA/YpW0k/src/compiler/execution.jl:100 [inlined]
 [13] iwrite_sendbufs!(n::Int64, dim::Int64, A::CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}, i::Int64)
    @ ImplicitGlobalGrid ~/.julia/packages/ImplicitGlobalGrid/b8fz5/src/update_halo.jl:347
 [14] _update_halo!(fields::CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer})
    @ ImplicitGlobalGrid ~/.julia/packages/ImplicitGlobalGrid/b8fz5/src/update_halo.jl:38
 [15] update_halo!(A::CUDA ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer})
    @ ImplicitGlobalGrid ~/.julia/packages/ImplicitGlobalGrid/b8fz5/src/update_halo.jl:27
 [16] res::error.jl:91
  [2] macro expansion
    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/error.jl:101 [inlined]
  [3] cuMemHostGetDevicePointer_v2(pdptr::Base.RefValue{CUDA.CuPtr{Nothing}}, p::Ptr{Nothing}, Flags::Int64)
    @ CUDA ~/.julia/packages/CUDA/YpW0k/lib/utils/call.jl:26
  [4] convert
    @ ~/.julia/packages/CUDA/YpW0k/lib/cudadrv/memory.jl:130 [inlined]
  [5] unsafe_convert
    @ ~/.julia/packages/CUDA/YpW0k/src/array.jl:320 [inlined]
  [6] pointer
    @ ~/.julia/packages/CUDA/YpW0k/src/array.jl:275 [inlined]
  [7] unsafe_convert
    @ ~/.julia/packages/CUDA/YpW0k/src/array.jl:327 [inlined]
  [8] adapt_structure
    @ ~/.julia/packages/CUDA/YpW0k/src/compiler/execution.jl:139 [inlined]
  [9] adapt
    @ ~/.julia/packages/Adapt/RGNRk/src/Adapt.jl:40 [inlined]
 [10] cudaconvert(arg::CUDA.CuArray{Float64, 1, CUDA.Mem.HostBuffer})
    @ CUDA ~/.julia/packages/CUDA/YpW0k/src/compiler/execution.jl:152
 [11] map
    @ ./tuple.jl:216 [inlined]
 [12] macro expansion
    @ ~/.julia/packages/CUDA/YpW0k/src/compiler/execution.jl:100 [inlined]
 [13] iwrite_sendbufs!(n::Int64, dim::Int64, A::CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer}, i::Int64)
    @ ImplicitGlobalGrid ~/.julia/packages/ImplicitGlobalGrid/b8fz5/src/update_halo.jl:347
 [14] _update_halo!(fields::CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer})
    @ ImplicitGlobalGrid ~/.julia/packages/ImplicitGlobalGrid/b8fz5/src/update_halo.jl:38
 [15] update_halo!(A::CUDA.CuArray{Float64, 2, CUDA.Mem.DeviceBuffer})
    @ ImplicitGlobalGrid ~/.julia/packages/ImplicitGlobalGrid/b8fz5/src/update_halo.jl:27
 [16] diffusion_2D(; do_visu::Bool)
    @ Main /scratch/lraess/course-101-0250-00/lecture08_teachers/diffusion_2D_perf_multixpu.jl:62
 [17] top-level scope
    @ /scratch/lraess/course-101-0250-00/lecture08_teachers/diffusion_2D_perf_multixpu.jl:84
in expression starting at /scratch/lraess/course-101-0250-00/lecture08_teachers/diffusion_2D_perf_multixpu.jl:84
diffusion_2D(; do_visu::Bool)
    @ Main /scratch/lraess/course-101-0250-00/lecture08_teachers/diffusion_2D_perf_multixpu.jl:62
 [17] top-level scope
    @ /scratch/lraess/course-101-0250-00/lecture08_teachers/diffusion_2D_perf_multixpu.jl:84
in expression starting at /scratch/lraess/course-101-0250-00/lecture08_teachers/diffusion_2D_perf_multixpu.jl:84
```

**Comment by omlins** [comment]
This is likely an issue with CUDA.jl. A [somewhat similar issue](https://github.com/JuliaGPU/CUDA.jl/issues/1125) was fixed recently in CUDA.jl.
As you noted, downgrading CUDA to v3.3.6 is currently required.
I will update the Compat ASAP and make release.

**Comment by luraess** [comment]
Using the following MWE, the issue could be solved in CUDA.jl with PR https://github.com/JuliaGPU/CUDA.jl/pull/1342
```julia
using CUDA
function register(buf::Array{T}) where T <: Number
    rbuf = Mem.register(Mem.Host, pointer(buf), sizeof(buf), Mem.HOSTREGISTER_DEVICEMAP);
    rbuf_d = convert(CuPtr{T}, rbuf);
    return unsafe_wrap(CuArray, rbuf_d, size(buf)), rbuf;
end
A = rand(2,2)
register(A)
```

**Comment by omlins** [comment]
Fixed in #32
